File: 0101-0200/0110.Balanced_Binary_Tree.cpp
//0MS

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    int findDepth(TreeNode* &root, bool &isBalanced) {
        // bc
        if (!root) return 0;

        int left = findDepth(root->left, isBalanced);
        int right = findDepth(root->right, isBalanced);
        // check for curr node is balanced
        if (isBalanced && abs(left - right) > 1) isBalanced = false;
        
        return max(left, right) + 1;
    }

    bool isBalanced(TreeNode* root) {
        // // 01 approach - o(n2)
        // // bc
        // if (!root) return true;

        // // left subtree
        // bool op3 = (abs(findDepth(root->left)-findDepth(root->right)) <= 1) ? true : false;
        // bool op1 = isBalanced(root->left);
        // bool op2 = isBalanced(root->right);

        // return (op1 && op2 && op3);

        // 02 approach - o(n)
        bool isBalanced = true;
        findDepth(root, isBalanced);
        return isBalanced;
    }
};

File: 0101-0200/0110.Balanced_Binary_Tree.js
//1ms

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function (root) {
    const dfs = (root) => {
        if (!root) {
            return 0
        }
        let l = dfs(root.left)
        if(l === -1) {
            return -1
        }
        let r = dfs(root.right)
        if(r === -1) {
            return -1
        }
        let diff = Math.abs(l - r)
        if (diff <= 1) {
            return Math.max(l, r) + 1
        } else {
            return -1
        }
    }
    return dfs(root) !== -1
};
